modules/data-streams/src/main/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,191 @@
+import org.elasticsearch.index.IndexMode;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.index.mapper.DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER;
+
+/**
+ * A component that updates the 'index.time_series.end_time' index setting for the most recently added backing index of a tsdb data stream.
+ */
+public class UpdateTimeSeriesRangeService extends AbstractLifecycleComponent implements LocalNodeMasterListener {
+
+    private static final Logger LOGGER = LogManager.getLogger(UpdateTimeSeriesRangeService.class);
+
+    private final ThreadPool threadPool;
+    private final ClusterService clusterService;
+
+    private volatile TimeValue pollInterval;
+    private volatile Scheduler.Cancellable job;
+    private final AtomicBoolean running = new AtomicBoolean(false);
+
+    UpdateTimeSeriesRangeService(Settings settings, ThreadPool threadPool, ClusterService clusterService) {
+        this.pollInterval = DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.get(settings);
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL, this::setPollInterval);
+    }
+
+    void perform(Runnable onComplete) {
+        if (running.compareAndSet(false, true)) {
+            LOGGER.debug("starting tsdb update task");
+            clusterService.submitStateUpdateTask("update_tsdb_data_stream_end_times", new ClusterStateUpdateTask(Priority.URGENT) {
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    return updateTimeSeriesTemporalRange(currentState, Instant.now());
+                }
+
+                @Override
+                public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                    running.set(false);
+                    onComplete.run();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    running.set(false);
+                    LOGGER.warn("failed to update tsdb data stream end times", e);
+                    onComplete.run();
+                }
+
+            }, ClusterStateTaskExecutor.unbatched());
+        } else {
+            LOGGER.debug("not starting tsdb update task, because another execution is still running");
+        }
+    }
+
+    void setPollInterval(TimeValue newValue) {
+        LOGGER.info(
+            "updating [{}] setting from [{}] to [{}]",
+            DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.getKey(),
+            pollInterval,
+            newValue
+        );
+        this.pollInterval = newValue;
```

Comment:
I think this needs to unschedule and then reschedule the task, otherwise updating the poll interval just changes the variable without affecting the scheduled invocation

Reply:
implemented via: 6ee565e47a08a049b1beb83f0b15aa90ba7d578b